src/neorados/cls/sem_set.h

```diff
@@ -0,0 +1,305 @@
+  namespace buffer = ::ceph::buffer;
+  buffer::list in;
+  ss::increment call{std::move(keys)};
+  encode(call, in);
+  return ClsWriteOp{[in = std::move(in)](WriteOp& op) {
+    op.exec(ss::CLASS, ss::INCREMENT, in);
+  }};
+}
+
+/// \brief Increment semaphores
+///
+/// Append a call to a write operation that increments the semaphores
+/// on a set of keys.
+///
+/// \param keys Keys to increment
+///
+/// \return The ClsWriteOp to be passed to WriteOp::exec
+template<std::input_iterator I>
+[[nodiscard]] inline auto increment(I begin, I end)
+  requires std::is_convertible_v<typename I::value_type, std::string>
+{
+  namespace ss = ::cls::sem_set;
+  namespace buffer = ::ceph::buffer;
+  buffer::list in;
+  ss::increment call{begin, end};
+  encode(call, in);
+  return ClsWriteOp{[in = std::move(in)](WriteOp& op) {
+    op.exec(ss::CLASS, ss::INCREMENT, in);
+  }};
+}
+
+/// \brief Decrement semaphore
+///
+/// Append a call to a write operation that decrements the semaphore
+/// on a key.
+///
+/// \param key Key to decrement
+///
+/// \return The ClsWriteOp to be passed to WriteOp::exec
+[[nodiscard]] inline auto decrement(std::string key)
+{
+  namespace ss = ::cls::sem_set;
+  namespace buffer = ::ceph::buffer;
+  buffer::list in;
+  ss::decrement call{std::move(key)};
+  encode(call, in);
+  return ClsWriteOp{[in = std::move(in)](WriteOp& op) {
+    op.exec(ss::CLASS, ss::DECREMENT, in);
+  }};
+}
+
+/// \brief Decrement semaphores
+///
+/// Append a call to a write operation that decrements the semaphores
+/// on a set of keys.
+///
+/// \param keys Keys to decrement
+///
+/// \return The ClsWriteOp to be passed to WriteOp::exec
+[[nodiscard]] inline auto decrement(std::initializer_list<std::string> keys)
+{
+  namespace ss = ::cls::sem_set;
+  namespace buffer = ::ceph::buffer;
+  buffer::list in;
+  ss::decrement call{keys};
+  encode(call, in);
+  return ClsWriteOp{[in = std::move(in)](WriteOp& op) {
+    op.exec(ss::CLASS, ss::DECREMENT, in);
+  }};
+}
+
+/// \brief Decrement semaphores
+///
+/// Append a call to a write operation that decrements the semaphores
+/// on a set of keys.
+///
+/// \param keys Keys to decrement
+///
+/// \return The ClsWriteOp to be passed to WriteOp::exec
+[[nodiscard]] inline auto decrement(std::unordered_set<std::string> keys)
+{
+  namespace ss = ::cls::sem_set;
+  namespace buffer = ::ceph::buffer;
+  buffer::list in;
+  ss::decrement call{std::move(keys)};
+  encode(call, in);
+  return ClsWriteOp{[in = std::move(in)](WriteOp& op) {
+    op.exec(ss::CLASS, ss::DECREMENT, in);
+  }};
+}
+
+/// \brief decrement semaphores
+///
+/// Append a call to a write operation that decrements the semaphores
+/// on a set of keys.
+///
+/// \param keys Keys to decrement
+///
+/// \return The ClsWriteOp to be passed to WriteOp::exec
+template<std::input_iterator I>
+[[nodiscard]] inline auto decrement(I begin, I end)
+  requires std::is_convertible_v<typename I::value_type, std::string>
```

Comment:
`I::value_type` should probably be `std::iterator_traits<I>::value_type` so it can work with arrays, whose iterators are raw pointers

Reply:
Thank you.